pkg/dns/libvirt/provider.go

```diff
@@ -0,0 +1,109 @@
+package libvirt
+
+import (
+	"encoding/xml"
+	"net"
+	"strings"
+
+	libvirt "github.com/libvirt/libvirt-go"
+
+	configv1 "github.com/openshift/api/config/v1"
+	iov1 "github.com/openshift/cluster-ingress-operator/pkg/api/v1"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+)
+
+var (
+	// _ dns.Provider = &Provider{}
+	log = logf.Logger.WithName("entrypoint")
+)
+
+type Provider struct {
+	// config is required input.
+	config Config
+}
+
+type Config struct {
+	Cluster string
+	Domain  string
+}
+type Host struct {
+	XMLName xml.Name `xml:"host"`
+	Ip      string   `xml:"ip,attr"`
+	// DnsRecord []hostname `xml:"hostname"`
+	DnsRecord string `xml:"hostname"`
+}
+
+type action = libvirt.NetworkUpdateCommand
+
+const (
+	upsertAction action = 3
+	deleteAction action = 2
+)
+
+// New creates (but does not start) a new operator from configuration.
+func New(config Config) (*Provider, error) {
+	provider := &Provider{
+		config: config,
+	}
+	return provider, nil
+}
+
+func (p *Provider) Ensure(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	return p.Change(record, zone, upsertAction)
+}
+
+func (p *Provider) Delete(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	return p.Change(record, zone, deleteAction)
+}
+
+// Change methods will perform an action on a record.
+func (p *Provider) Change(record *iov1.DNSRecord, zone configv1.DNSZone, action action) error {
+	// Create a new connections to qemu
+	conn, err := libvirt.NewConnect("qemu:///system")
```

Comment:
The connection URL should not be hardcoded. You'll need to get this somehow from the Installer, just like libvirt provider does: https://github.com/openshift/cluster-api-provider-libvirt/blob/master/pkg/apis/libvirtproviderconfig/v1beta1/libvirtmachineproviderconfig_types.go#L20